src/data-workspace/display/display.js

```diff
@@ -0,0 +1,183 @@
+import { useConfig } from '@dhis2/app-runtime'
+import i18n from '@dhis2/d2-i18n'
+import { Button, IconLaunch16 } from '@dhis2/ui'
+import PropTypes from 'prop-types'
+import React from 'react'
+import { useExchangeContext } from '../../exchange-context/index.js'
+import { Warning } from '../../shared/index.js'
+import { Table } from '../table/index.js'
+import styles from './display.module.css'
+
+// TBD re: data conversion for formatting
+
+const convertToObjectFormat = ({
+    data,
+    dx_index,
+    ou_index,
+    pe_index,
+    value_index,
+}) => {
+    const objectFormat = {}
+
+    for (const row of data?.rows) {
+        if (
+            !Object.prototype.hasOwnProperty.call(objectFormat, row[ou_index])
+        ) {
+            objectFormat[row[ou_index]] = {}
+        }
+        if (
+            !Object.prototype.hasOwnProperty.call(
+                objectFormat[row[ou_index]],
+                row[dx_index]
+            )
+        ) {
+            objectFormat[row[ou_index]][row[dx_index]] = {}
+        }
+        if (
+            !Object.prototype.hasOwnProperty.call(
+                objectFormat[row[ou_index]][row[dx_index]],
+                row[pe_index]
+            )
+        ) {
+            objectFormat[row[ou_index]][row[dx_index]][row[pe_index]] = {}
+        }
+        objectFormat[row[ou_index]][row[dx_index]][row[pe_index]] =
+            row[value_index]
```

Comment:
The code I have highlighted here is mostly an extreme example of what I don't like about this file in general. There are way too many nested `[]` selectors (and also way too many conditional chaining operators `?.`). I'm sure you have tested this and it is all is working correctly, but I personally would not dare to implement any changes here and would probably end up rewriting things until I got to a point where I can actually understand what's going on by looking at the code. In the process I would probably break something 😓 and cause a regression. Before we merge this PR, I think we need to:
- Clean up the implementation so it is more readable. 
- Quite likely things are going to remain fairly complex even after cleanup, so it'd be good to also add a unit test for this, so we catch regressions quickly

Let me know if you want to discuss potential ways of cleaning this up.

Reply:
I've update the implementation to keep this conversion to an objectFormat, but have cleaned that up (https://github.com/dhis2/data-exchange-app/commit/d612cf7e80fdb02f0c0f6fce3e180f0c7b13ba88) so it's hopefully more readable. 

Will add unit tests for `convertToObjectFormat` and `formatData` functions

Reply:
Have now added [tests](https://github.com/dhis2/data-exchange-app/blob/417dacc5f8f40bda7e72c9fba8a8ddeac50a091e/src/data-workspace/requests-display/requests-display.test.js) for the refactor

Reply:
💪 well done. Thanks for this.